runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java

```diff
@@ -163,7 +167,7 @@ public Node apply(MutableNetwork<Node, Edge> input) {
       componentsBuilder.putEnvironments(envId, Environments.JAVA_SDK_HARNESS_ENVIRONMENT);
     }
 
-    // Use default WindowingStrategy as the fake one.
+    // Use default WindowingStrategy as the fake one. But it may be overrided later if it's a FixedWindow.
     // TODO: should get real WindowingStategy from pipeline proto.
```

Comment:
I agree and getting pipeline proto from Dataflow service should be an ultimate solution. But it's kind of hard for me to change the proto from Dataflow service, and the proto changes will also take some time. This generated WindowingStrategy is a short term workaround for now. We are discussing the possibility of changing proto from Dataflow service.

Reply:
So the v1beta3 protos expose neither the windowing strategy nor a reference that can be used to tie back to the original PCollection? Unfortunate...

In this case, let's add a (possibly memoized) method `getWindowingStrategyForCoder(Coder<?> windowCoder, componentsBuilder)` rather than creating and injecting a default and then later possibly overwriting it. It can have a case for Global and Interval windows, and throw an error otherwise.